lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskStatus.cs

```diff
@@ -98,28 +98,48 @@ public void SetException(Exception e)
         {
             lock (_heartBeatManager)
             {
-                try
+                if (_lastException.IsPresent())
                 {
-                    if (!_lastException.IsPresent())
-                    {
-                        _lastException = Optional<Exception>.Of(e);
-                    }
+                    return;
+                }
 
-                    State = TaskState.Failed;
+                _lastException = Optional<Exception>.Of(e);
+                State = TaskState.Failed;
+
+                try
+                {
                     _taskLifeCycle.Stop();
                 }
-                finally
+                catch (Exception stopException)
                 {
-                    Heartbeat();
+                    Utilities.Diagnostics.Exceptions.Caught(
+                        stopException, Level.Error, "Exception thrown on Task Stop", LOGGER);
                 }
+
+                Heartbeat();
             }
         }
 
         public void SetResult(byte[] result)
         {
             lock (_heartBeatManager)
             {
+                if (_state == TaskState.Failed)
+                {
+                    if (!_lastException.IsPresent())
+                    {
+                        throw new ApplicationException("Expected failed Task to already have set an Exception.");
```

Comment:
We only reach `_state = Failed` in `SetException`, and `SetException` has to set `_lastException` to a "present" `Optional`. Everything is also done with a lock held on `_heartbeatManager`. Thus, if we ever reach a state where `_state == Failed` but `_lastException` is not "present", we must have had some sort of application logic and we should not call `SetException` again. Instead, we should crash the Evaluator with an `ApplicationException` that is not caught in the upper layers.